### PR TITLE
Adjust Grpc.Core to build for Unity (iOS + Android)

### PR DESCRIPTION
**Pull request by zaneclaes**
@jtattermusch here it is! I've updated this to support both iOS and Android. I'll leave some notes as comments on the code.

Approach
----
I'm really not happy with my approach here, to be honest. You'll note I had to create two huge copy/pasta classes, `LibBridgeExt` and `LibBridgeInternal` for the `DllImport("grpc_csharp_ext")` syntax and `DllImport("__Internal")` syntax. I'm not sure there's a way around that, but at least I wish I could use reflection in `NativeMethods` to simplify the loading process; unfortunately `GetMethods` doesn't work well with `static` classes.

At its core, the biggest challenge comes down to the fact that I do not have access to Unity's defines. If I did, I could compile out/in the correct `pluginName`. Instead, I end up adding this to my Unity project...
```
    /// <summary>
    /// Use the Application.platform to detect which kind of GRPC library should be loaded.
    /// </summary>
    private void ConfigureGrpcLibBridgeType() {
      switch (Application.platform) {
        case RuntimePlatform.Android: LibBridge.type = LibBridgeType.Ext; break;
        case RuntimePlatform.IPhonePlayer: LibBridge.type = LibBridgeType.Internal; break;
        case RuntimePlatform.tvOS: LibBridge.type = LibBridgeType.Internal; break;
        default: LibBridge.type = LibBridgeType.UnmanagedLibrary; break;
      }
    }
```
I could equally have used the defines, like `#if UNITY_IOS ... #end`. See the full list here: https://docs.unity3d.com/Manual/PlatformDependentCompilation.html I didn't use this method in my game, because the code which configures GRPC actually lives in a DLL itself... which does not have access to the Unity defines, either. However, it does have the `UnityEngine` runtime, so it can run the above checks.

Tested so far on iOS and Android. Need to double-check that native platforms still work, and my dream would also to get WebAssembly working...

Testing
----
- [x] Android
- [x] iOS (iPhone)
- [x] Linux x86+x64
- [x] OSX
- [ ] Windows
- [ ] Web (WebAssembly)

### CODE REVIEW

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @zaneclaes isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

File: src/csharp/Grpc.Core/Internal/DefaultSslRootsOverride.cs [hunk 1]
```diff
@@ -41,16 +41,17 @@ public static void Override(NativeMethods native)
         {
             lock (staticLock)
             {
+                /*
```
Comment (zaneclaes) [inline src/csharp/Grpc.Core/Internal/DefaultSslRootsOverride.cs]: If I don't comment this out, I get:
```
IOException: Error loading the embedded resource "Grpc.Core.roots.pem"
Grpc.Core.Internal.DefaultSslRootsOverride.Override (Grpc.Core.Internal.NativeMethods native) (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.Internal.NativeExtension..ctor () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.Internal.NativeExtension.Get () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.Internal.NativeMethods.Get () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.GrpcEnvironment.GrpcNativeInit () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.GrpcEnvironment..ctor () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.GrpcEnvironment.AddRef () (at <f1624d4830b344a9bbc721e6917f94b1>:0)
Grpc.Core.Channel..ctor (System.String target, Grpc.Core.ChannelCredentials credentials, System.Collections.Generic.IEnumerable`1[T] options) (at <f1624d4830b344a9bbc721e6917f94b1>:0)
```

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @zaneclaes isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Comment by jtattermusch** [comment]
@zaneclaes   Thanks for sharing your progress!

File: src/core/lib/support/log_android.cc [hunk 2]
```diff
@@ -63,7 +63,9 @@ extern "C" void gpr_default_log(gpr_log_func_args* args) {
 
   asprintf(&output, "%s:%d] %s", display_file, args->line, args->message);
 
-  __android_log_write(severity_to_log_priority(args->severity), "GRPC", output);
+  // Do not write to Android b/c of compiler error in Unity
+  // The rewritten GRPC core native log redirector does not need this anyways.
+  // __android_log_write(severity_to_log_priority(args->severity), "GRPC", output);
```
Comment (zaneclaes) [inline src/core/lib/support/log_android.cc]: Actually, it'd be a runtime error that `__android_log_write` was undefined.

File: src/csharp/Grpc.Core/Internal/NativeLogRedirector.cs [hunk 3]
```diff
@@ -51,38 +52,41 @@ public static void Redirect(NativeMethods native)
...
+                    line,
                     Marshal.PtrToStringAnsi(msgPtr));
-                
+
                 switch (severityString)
                 {
                     case "D":
-                        logger.Debug(message);
+                        // logger.Debug(message);
+                        Logz.LogDebug(message, "GRPC");
```
Comment (zaneclaes) [inline src/csharp/Grpc.Core/Internal/NativeLogRedirector.cs]: This is my own logging tool, which I'd rather route everything through anyways. I suppose I could have conformed to the `ILogger` interface and assigned it to the `GrpcEnvironment.Logger` instead.

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 4]
```diff
@@ -18,165 +18,153 @@
...
-            // Look for all native library in all possible locations in given order.
-            string[] paths = new[] { classicPath, netCorePublishedAppStylePath, netCoreAppStylePath};
-            return new UnmanagedLibrary(paths);
-        }
+    /// <summary>
+    /// Detects which configuration of native extension to load and load it.
+    /// </summary>
+    private static UnmanagedLibrary Load() {
+      if (LibBridge.type != LibBridgeType.UnmanagedLibrary) {
+        return null;
```
Comment (zaneclaes) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: Apologies; my linter rather over-zealously edited this file. This should be the only line which is actually necessary.

**Comment by zaneclaes** [comment]
Yikes, I meant to push this against my fork. Oh well; this does equally well demonstrating the tweaks I had to make @jtattermusch ... one note: I did this all against `1.8.3` release, instead of master, which was the NuGet package at the time that I started working on this. Let me know how I can help get this made a part of master, though -- it'd be great to be able to stay on the latest GRPC version :)

**Comment by jtattermusch** [comment]
@zaneclaes   how do you package Grpc so it can be used in unity?
1. import all the Grpc.Core sources into a unity project and then create a ".unitypackage" from it (and you then use the .unitypackage in your actual unity projects).
2. some other way

**Comment by zaneclaes** [comment]
I actually have a somewhat unique situation.
1) `Omni` is an open-source network layer and data manager I'm working on. It carries no dependencies on Unity, but does support GRPC. Right now, I have all of `GrpcCore` just included into the `Omni` project as source code (copy/paste). The other libraries are loaded through NuGet as usual. This complies to a single, cross-platform `Omni.dll`.
2) `Moongate` is another library, this one specific to my game, that relies upon `Omni`. The reason it is a library is that it lets me share code between my client and server. This library has dependencies on `UnityEngine`, which lets me run the config code I posted in the PR description.

My Unity project, then, just has `Omni.dll`, `Moongate.dll`, and of course the necessary Protobuf and `grpc_csharp_ext` libraries all placed in the unity `Plugins` directory.

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 5]
```diff
@@ -74,32 +74,43 @@ public NativeMethods NativeMethods
...
+      var classicPath = Path.Combine(assemblyDirectory, GetNativeLibraryFilename());
+      // With dotnet cli project targeting netcoreapp1.0, projects will use Grpc.Core assembly directly in the location where it got restored
+      // by nuget. We locate the native libraries based on known structure of Grpc.Core nuget package.
+      // When "dotnet publish" is used, the runtimes directory is copied next to the published assemblies.
+      string runtimesDirectory = string.Format("runtimes/{0}/native", GetPlatformString());
+      var netCorePublishedAppStylePath = Path.Combine(assemblyDirectory, runtimesDirectory, GetNativeLibraryFilename());
+      var netCoreAppStylePath = Path.Combine(assemblyDirectory, "../..", runtimesDirectory, GetNativeLibraryFilename());
+
+      // e.g., on OSX the plugin lives at [APPNAME].app/Contents/Plugins/
+      var unityPluginsPath = Path.Combine(assemblyDirectory, "../..", "Plugins", GetNativeLibraryFilename());
```
Comment (zaneclaes) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: This is what lets GRPC know where to search for Unity plugins, which are not included where GRPC expects.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @zaneclaes isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 6]
```diff
@@ -174,7 +185,7 @@ private static string GetNativeLibraryFilename()
             }
             if (PlatformApis.IsMacOSX)
             {
-                return string.Format("libgrpc_csharp_ext.{0}.dylib", architecture);
+                return string.Format("libgrpc_csharp_ext.{0}.bundle", architecture);
```
Comment (zaneclaes) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs]: See this topic: https://answers.unity.com/questions/23615/how-to-make-unity-find-dylib-files.html

**Comment by jtattermusch** [comment]
@zaneclaes   as this PR is basically superseded by https://github.com/grpc/grpc/pull/14605, I will close it now, but we might still use it for future reference (so let's keep the branch with code around). Thanks for creating it and sharing your results, it was great that I could get inspired by your code and the approach you used!

**Pull request closed without merging**